docs(core): add docs for i18n tokens

ocombe opened the pull request:
## PR Type
What kind of change does this PR introduce?
```
[x] Documentation content changes
```

## What is the current behavior?
No API docs on i18n tokens

## What is the new behavior?
Documentation is available

## Does this PR introduce a breaking change?
```
[x] No
```

Note: needs to be merged with #20623

ocombe committed:
docs(core): add docs for i18n tokens

mary-poppins commented:
You can preview 975a5ae at https://pr17920-975a5ae.ngbuilds.io/.

mary-poppins commented:
You can preview 82b7a06 at https://pr17920-82b7a06.ngbuilds.io/.

vicb commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -9,21 +9,69 @@
 /**
+ * This token is used to determine the locale of the application. You have to use a valid locale
+ * id code, as explained [here](http://userguide.icu-project.org/locale), because the `LOCALE_ID`
+ * is used for i18n extraction, in i18n pipes ({@link DatePipe}, {@link I18nPluralPipe},
```
used "by" * 2(instead of in)

vicb commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -9,21 +9,69 @@
+ * {@link CurrencyPipe}, {@link DecimalPipe} and {@link PercentPipe}) and in ICU expressions
+ * (See [i18n guide](https://angular.io/guide/i18n)).
+ *
+ * This token has to be provided at bootstrap because it is used by the HTML compiler.
```
could you please describe JIT vs AOT

vicb commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -9,21 +9,69 @@
 /**
+ * Use this enum at bootstrap to define the strategy that the `CompilerConfig` should use in case
+ * of missing translations (See [i18n guide](https://angular.io/guide/i18n)):
+ * - Error
```
docs what each value means

vicb reviewed (changes requested):
We should probably make what apply to JIT only more explicit. Otherwise :+1:

mary-poppins commented:
You can preview ac67960 at https://pr17920-ac67960.ngbuilds.io/.

mary-poppins commented:
You can preview 9d3ce7c at https://pr17920-9d3ce7c.ngbuilds.io/.

ocombe commented:
@vicb let me know if the updates are ok for you

vicb commented:
blocked on https://github.com/angular/angular/issues/16960

vicb commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -9,21 +9,115 @@
 /**
+ * This token is used to determine the locale of the application. You have to use a valid locale
+ * id code, as explained [here](http://userguide.icu-project.org/locale), because the `LOCALE_ID`
+ * is used for i18n extraction, by i18n pipes ({@link DatePipe}, {@link I18nPluralPipe},
```
remove the link here (because it will get outdated at some point) and add the link there

vicb commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -9,21 +9,115 @@
+ * {@link CurrencyPipe}, {@link DecimalPipe} and {@link PercentPipe}) and by ICU expressions
+ * (see [i18n guide](https://angular.io/guide/i18n)).
+ *
+ * If you are using AOT, you can define the locale with the cli parameter `--locale` for ngc
```
should not be on this API doc - move to angular.io

ocombe replied:
something like: "If you are using AOT see [i18n guide](https://angular.io/guide/i18n)" ?

vicb commented on packages/core/src/i18n/tokens.ts:
````diff
@@ -9,21 +9,115 @@
+ *
+ * ### Example
+ *
+ * ```typescript
````
should be on angular.io (we do not use inline exemple any more in the API doc any more but include them from the examples folder AFAIR)

vicb replied:
The comment on this doc block apply to all blocks in this PR

mary-poppins commented:
You can preview 2d54ba1 at https://pr17920-2d54ba1.ngbuilds.io/.

mary-poppins commented:
You can preview 12b706b at https://pr17920-12b706b.ngbuilds.io/.

vicb commented:
As commented earlier this was not merged because it does not work as documented in JIT ("blocked on #16960").

Please either fix the root cause or remove the misleading doc from this PR.

I would like this PR closed by eow. Either by merging it or closing it as non applicable.

ocombe commented:
I've updated the docs, waiting for the docs team to give me my approval on the changes. I've noticed that we removed the docs on the missing translation strategy from the guide in v5, which is why I added it again here at the same time.

mary-poppins commented:
You can preview bb24d4f at https://pr17920-bb24d4f.ngbuilds.io/.

mary-poppins commented:
You can preview 73036fe at https://pr17920-73036fe.ngbuilds.io/.

mary-poppins commented:
You can preview 7951e7c at https://pr17920-7951e7c.ngbuilds.io/.

mary-poppins commented:
You can preview 203f5ce at https://pr17920-203f5ce.ngbuilds.io/.

vicb commented on aio/content/guide/i18n.md:
```diff
@@ -62,6 +61,23 @@ For more information about Unicode locale identifiers, see the
 [the Angular repository](https://github.com/angular/angular/tree/master/packages/common/locales).
 
+The locale identifiers used by CLDR and Angular are based on [BCP47](http://www.rfc-editor.org/rfc/bcp/bcp47.txt)
+but since specifications change over time, here is a list of equivalences between old locale
```
Need some love from a native speaker.

jbogarthyde replied:
The locale identifiers used by CLDR and Angular are based on [BCP47](http://www.rfc-editor.org/rfc/bcp/bcp47.txt).  These specifications change over time; the following table maps old identifiers to current ones at time of writing.

jenniferfell replied:
Agree with Judy's recommendation.

jenniferfell replied:
(optional) Additional refinement: Can we say what we mean by "old"? Angular v4? BCP47 in 2015?

ocombe replied:
old = anything before the last update of BCP47

jenniferfell replied:
Can we be specific about a date or version of BCP47 that is "old"? The standard I found is from 2009, so is old "prior to the 2009 update of BCP47"? https://tools.ietf.org/html/bcp47

ocombe replied:
I'm not sure about the date of each change, let's keep it like this for now.

vicb commented:
LGTM for the correctness of the information.
Needs some proof reading / updates by someone on the doc team

jbogarthyde commented on aio/content/guide/i18n.md:
```diff
@@ -544,7 +551,6 @@ This sample file is easy to translate without a special editor or knowledge of F
 </div>
```
Notice that the tool has assigned generated IDs, which are hard to work with and are regenerated when either the text or meaning changes. We recommend that you [specify custom IDs](guide/i18n#custom-id), in order to avoid [maintenance](guide/i18n#maintenance) problems.

jbogarthyde commented on aio/content/guide/i18n.md:
```diff
@@ -598,7 +602,6 @@ Here they are together, after translation:
-
 ### Translate a nested expression
 
 A nested expression is similar to the previous examples. As in the previous example, there are 
```
A nested expression, like a select statement, gets broken into separate translation units for the outer and inner expressions. The first one contains...

jbogarthyde commented on aio/content/guide/i18n.md:
```diff
@@ -621,7 +624,6 @@ The entire template translation is complete. The next section describes how to l
 into the app.
```
Lose this line - it doesn't add anything and there is another subsection before the "next section".

jbogarthyde commented on aio/content/guide/i18n.md:
```diff
@@ -621,7 +624,6 @@ The entire template translation is complete. The next section describes how to l
-
 ### The app and its translation file
 
 The sample app and its translation file are now as follows:
```
After completing the translations, the sample app and translation files are ready to be merged. The prepared files look like this:

jbogarthyde commented on aio/content/guide/i18n.md:
```diff
@@ -713,3 +712,26 @@ Then provide the `LOCALE_ID` in the main module:
 <code-example path="i18n/doc-files/app.module.ts" title="src/app/app.module.ts" linenums="false">
 </code-example>
+
+
```
There are a bunch of subsections missing here that show up in Mary Poppins -- I wanted to make an edit in the first one, "SystemsJS text plugin" -- is it part of another PR?
The others are "Create translation providers", "Bootstrap with translation providers", and "Internationalization with the AOT compiler" -- do you want a copy edit on those?

jbogarthyde commented on aio/content/guide/i18n.md:
```diff
@@ -713,3 +712,26 @@ Then provide the `LOCALE_ID` in the main module:
+bootstrap:
+
+<code-example path="i18n/doc-files/main.3.ts" title="src/main.ts">
+</code-example>
```
The last section is also missing -- and there is a reference to it earlier: guide/i18n#file-maintenance-and-id-changes

jbogarthyde commented on aio/content/guide/i18n.md:
```diff
@@ -62,6 +61,23 @@ For more information about Unicode locale identifiers, see the
+| Chinese Traditional Hong Kong | zh-hk             | zh-Hant-HK    |
+
 
 ## i18n pipes
```
This section is missing from the Mary Poppins page I'm looking at (https://pr17920-9d3ce7c.ngbuilds.io/guide/i18n)
The descriptive text for the code example can use some work, but I can't do it without seeing the code that's being described. 
Did the pipes section get moved somewhere else? It does seem to be kind of out-of-place here.

jbogarthyde reviewed (commented):
Copy-edit suggestions

ocombe commented:
@jbogarthyde use the last preview link: https://pr17920-203f5ce.ngbuilds.io/guide/i18n
I've removed the other ones to avoid confusion

jenniferfell commented on aio/content/guide/i18n.md:
```diff
@@ -78,6 +94,11 @@ If you want to import locale data for other languages, you can do it manually:
 </code-example>
 
+The first parameter is an object containing the locale data imported from `@angular/common/locales`
+while the second parameter is the locale id for which you want to register those data. This second
```
Recommend shorter sentences: "The first parameter is an object containing the locale data imported from 'angular/common/locales'.

jenniferfell commented on aio/content/guide/i18n.md:
```diff
@@ -78,6 +94,11 @@ If you want to import locale data for other languages, you can do it manually:
+The first parameter is an object containing the locale data imported from `@angular/common/locales`
+while the second parameter is the locale id for which you want to register those data. This second
+parameter is optional but it can be used to register locale data for a custom locale id that is not
+in the list of locales supported by Angular.
```
Tech questions: I'm a bit confused because it sounds like the second parameter is used to associate imported locale data with a custom locale id, but the example shows the id 'fr' which is a standard predefined id. In this case, does the imported locale data get registered with that standard id, replacing any standard or previous registered data? If I don't include the second parameter, what happens with the object that represents the imported locale data? Do I  just have to use the object, because there is no shorthand (id) assigned?

ocombe replied:
if you don't specify the second parameter, it will use what's defined in the data (for "fr" it will be "fr", for en-GB it will be "en-GB"...)
But if you want to use "fr-FR" for fr data, you have to specify that, otherwise it'll use "fr" by default

jenniferfell replied:
Ah ha! That makes sense. So maybe change the example to "fr-FR" so it's not already using the default. Then lets add that info:
"The first parameter is an object containing the locale data imported from 'angular/common/locales'. By default, the imported locale data is registered with the locale id that is defined in the Angular locale data itself.  If you want to register the imported locale data with another locale id, use the second parameter to specify a custom locale id. For example, Angular's locale data defines the locale id for French as "fr". You can use the second parameter to associate the imported French locale data with the custom locale id "fr-FR."

jenniferfell commented on aio/content/guide/i18n.md:
```diff
@@ -656,12 +657,11 @@ format that Angular understands, such as `.xtb`.
-  * With [JIT](guide/i18n#jit), you provide the information at bootstrap time.
-
+  * With [AOT](guide/i18n#merge-aot), you pass the information as a CLI parameter.
+  * With [JIT](guide/i18n#merge-jit), you provide the information at bootstrap time.
```
Confirmed. Links work as desired.

jenniferfell commented on aio/content/guide/i18n.md:
```diff
@@ -713,3 +712,26 @@ Then provide the `LOCALE_ID` in the main module:
+
+{@a missing-translation}
+### Report missing translations
+When a translation is missing, the build will succeed but it will print a warning that you might
```
Edit: When a translation is missing, the build succeeds but generates a warning. You can configure...

jenniferfell commented on aio/content/guide/i18n.md:
```diff
@@ -713,3 +712,26 @@ Then provide the `LOCALE_ID` in the main module:
+When a translation is missing, the build will succeed but it will print a warning that you might
+easily overlook. You can configure the Angular compiler for different "missing translation" behaviors:
+
+* Error: throw if you have missing translations.
```
Edits:
Error: throw an error. 
Warning (default): Show a warning in the console or shell.
Ignore: do nothing.

jenniferfell replied:
We can add more detail to this list. 
Error: throws an error. If you are using AOT compilation, the build will fail. If you are using JIT compilation, the app will fail to load. 
Warning (default): show a 'Missing translation' warning in the console or shell.
Ignore: do nothing

jenniferfell commented on aio/content/guide/i18n.md:
```diff
@@ -713,3 +712,26 @@ Then provide the `LOCALE_ID` in the main module:
+easily overlook. You can configure the Angular compiler for different "missing translation" behaviors:
+
+* Error: throw if you have missing translations.
+* Warning (default): show a warning in the console and/or shell.
```
Can we offer more information? 
Error: Does the build still succeed or does it fail? What is the error message or keyword to search for in error logs or messages?
Warning: What is the warning or keyword to search for?

ocombe replied:
An error will fail the build for AOT, or fail the app loading for JIT.
The error/warning message will be something like that: `Missing translation for message "foo"`.

jenniferfell replied:
Let's be clear that's the default. 
By default, when a translation is missing, the build succeeds but generates a warning such as Missing translation for message "foo". You can configure the level of warning that is generated by the Angular compiler.

jenniferfell replied:
See above for suggestions to incorporate more of this detail into the bulleted list

jenniferfell commented on aio/content/guide/i18n.md:
```diff
@@ -713,3 +712,26 @@ Then provide the `LOCALE_ID` in the main module:
+* Warning (default): show a warning in the console and/or shell.
+* Ignore: do nothing.
+
+To change this behavior use the CLI parameter `--missingTranslation` with the name of the strategy
```
Edit: If you use the AOT compiler, you can specify the warning level by using the CLI parameter '--missingTranslation'. The example below shows how to set the warning level to error:

jenniferfell commented on aio/content/guide/i18n.md:
```diff
@@ -713,3 +712,26 @@ Then provide the `LOCALE_ID` in the main module:
+  ng serve --aot --missingTranslation=error
+</code-example>
+
+If you use JIT, you need to provide the missing translation strategy in the compiler config at
```
Edit: If you use the JIT compiler, specify the warning level in the compiler config at bootstrap by adding the 'MissingTranslationStrategy' parameter. The example below shows how to set the warning level to error:

(I'm not sure if the right word is parameter or option or setting. Choose the right word for how you talk about the compiler config file things.)

ocombe replied:
In this case the correct word would be "property"

jenniferfell commented:
Review completed. I understand that any comments we made on lines that aren't affected by this PR won't be incorporated at this time.

jenniferfell commented:
@ocombe Replied to your questions.

mary-poppins commented:
You can preview d7ec1c6 at https://pr17920-d7ec1c6.ngbuilds.io/.

gkalpak commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -9,21 +9,95 @@
+ *
+ * platformBrowserDynamic().bootstrapModule(AppModule, {
+ *   providers: [{provide: LOCALE_ID, useValue: 'en-US' }]
+ * });
```
Hard-coded example 😱 
<sub>(Ideally all code snippets should come from actual examples.)</sub>

gkalpak reviewed (approved):
LGTM (as long as both Travis and @jenniferfell are happy :smiley:)

jenniferfell commented:
I approve all the text changes. I'm okay with deferring work to fix the hard-coded example. 

Footnote on my approval:  I can't speak to why Travis is failing. That probably needs to be addressed before merge.
